base/utils/R/read.output.R

```diff
@@ -16,160 +16,241 @@
+    # explicitly.
+    ncfiles_sub <- basename(ncfiles)
+  }
   nc_years <- as.numeric(gsub("^(-?[[:digit:]]{4})\\.nc", "\\1", ncfiles_sub))
-  
-  if(!is.na(start.year) && !is.na(end.year)){
+
+  if (!is.na(start.year) && !is.na(end.year)) {
     if (lubridate::is.instant(start.year)) { # true if a Date, POSIXct, or POSIXlt
-      start.year = lubridate::year(start.year)
+      start.year <- lubridate::year(start.year)
     } else if (is.character(start.year)) {
-      start.year = as.numeric(start.year)
+      start.year <- as.numeric(start.year)
     } else {
       PEcAn.logger::logger.error("start.year must be of type numeric, character, Date, or POSIXt")
     }
     if (lubridate::is.instant(end.year)) {
-      end.year = lubridate::year(end.year)
+      end.year <- lubridate::year(end.year)
     } else if (is.character(end.year)) {
-      end.year = as.numeric(end.year)
+      end.year <- as.numeric(end.year)
     } else {
       PEcAn.logger::logger.error("end.year must be of type numeric, character, Date, or POSIXt")
     }
 
     # select only those *.nc years requested by user
     keep <- which(nc_years >= start.year & nc_years <= end.year)
     ncfiles <- ncfiles[keep]
-  } else if(length(nc_years) != 0){
-      PEcAn.logger::logger.info("No start or end year provided; reading output for all years")
-      start.year <- min(nc_years)
-      end.year <- max(nc_years)
+  } else if (length(nc_years) != 0) {
+    PEcAn.logger::logger.info(
+      "No start or end year provided; reading output for all years"
+    )
+    start.year <- min(nc_years)
+    end.year <- max(nc_years)
   }
 
-  years <- start.year:end.year
+  years <- seq(start.year, end.year)
   run_origin <- paste0(start.year, "-01-01")
 
   # throw error if no *.nc files selected/availible
   nofiles <- FALSE
   if (length(ncfiles) == 0) {
-    PEcAn.logger::logger.warn("read.output: no netCDF files of model output present for runid = ", 
-                runid, " in ", outdir, " for years ", start.year, ":", end.year, ". will return NA")
+    PEcAn.logger::logger.warn(
+      "read.output: no netCDF files of model output present",
+      "for runid = ", runid,
+      " in ", outdir,
+      " for years ", start.year, ":", end.year, ".",
+      "Returning NA."
+    )
     if (length(nc_years) > 0) {
-      PEcAn.logger::logger.info("netCDF files for other years present", nc_years)
+      PEcAn.logger::logger.info(
+        "netCDF files for other years present: ",
+        nc_years
+      )
     }
     nofiles <- TRUE
   } else {
-    PEcAn.logger::logger.info("Reading output for Years: ", start.year, " - ", end.year, 
-                "in directory:", outdir,
-                "including files", basename(ncfiles))
+    PEcAn.logger::logger.info(
+      "Reading output for Years: ", start.year, " - ", end.year,
+      "in directory:", outdir, "including files", basename(ncfiles)
+    )
   }
-  
+
   result <- list()
 
-  if (!nofiles) {
+  if (nofiles) {
```

Comment:
I see that this should be logically identical, but: Does the existing test coverage already check this section well? If not, it'd be nice to fix that so we can verify that this doesn't change anything unintended.

Reply:
Haven't looked at test coverage, though I definitely agree it's worthwhile. I won't have time to look into it this before AGU, though (but I may have some time at boring sessions _during_ AGU).

Reply:
I still _really_ wish we could test for `logger.*` messages with `expect_*` functions. Depending on exactly how `logger` works, I suppose it may be possible to use `capture.output` and then `grep` the underlying string. Now that I think of it, a `PEcAn.logger::expect_logger` function that defines a custom `testthat` expectation could be _really_ useful.

Reply:
I made a start at an `expect_log` in https://github.com/PecanProject/pecan/blob/develop/modules/data.atmosphere/tests/testthat/helper.R#L17 as kind of a one-off, and actually I think at the time you suggested moving somewhere more generally accessible... should we move that into PEcAn.logger?

Reply:
I think my excuses for not moving it earlier were that I  (1) didn't want to hardwire an `rlang` dependency into PEcAn.logger, and (2) was hoping maybe we'd "soon" change logger to produce message objects and therefore not need a special `expect_log`. 

With some care we can probably avoid point 1, and point 2 still sounds nice but isn't high on my list.

Reply:
Not pretty, but at least expect_output() should work with the Logger functions.

Reply:
@robkooper Do you have a demo of `expect_output` working with logger functions? I've never had any luck:

```
> testthat::expect_output(PEcAn.logger::logger.info("foo"), "foo")
2018-12-07 14:31:01 INFO   [eval_bare] : foo 
Error: `PEcAn.logger::logger.info("foo")` produced no output
```

Reply:
From approximately 5 seconds of testing (so may be wrong, or at least incomplete), I think this is because `testthat::expect_output` looks at `stdout`, but `PEcAn.logger::logger.*` prints to `stderr`. For instance, this fails...

```r
testthat::expect_output(message("hello"), "hello")
```

However, `testthat::expect_message(PEcAn.logger::logger.info("hello"), "hello")` _also_ fails, because, as @infotroph pointed out a while ago, the `PEcAn.logger::logger.*` functions don't signal a `message` [condition](https://adv-r.hadley.nz/conditions.html). 

This successfully captures the output as a string:

```r
out <- capture.output(PEcAn.logger::logger.info("hello"), type = "message")
```

So an eventual version of `expect_logger` could use this combined with `grepl`.

Reply:
In a686e965543d03042762d2da991acf39373a4d37, I added a bunch of unit tests to cover all of these changes.

In those tests, note the use of the `capture.output` + `expect_match`, which effectively serves the purpose of an `expect_log` function (and, without `expect_match`, can be used to prevent all of the verbose logging from polluting the `testthat` -- and therefore Travis! -- output). Note that my version is completely base R -- no `rlang` dependency necessary! We can port it to a `PEcAn.logger::expect_log` function at some point, but I'm gonna punt on that till a later PR.